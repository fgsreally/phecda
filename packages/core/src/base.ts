import { Empty, Unmount } from './decorators'
import { getTag } from './helper'
import { Events } from './types'

@Empty
export abstract class Base {
  private readonly __UNMOUNT_SYMBOL__: (() => void)[] = []
  private readonly __PROMISE_SYMBOL__: Promise<void> // for init

  abstract emitter: any
  constructor() {

  }

  get tag(): PropertyKey {
    return getTag(this)
  }

  then(cb: () => void, reject?: (e: any) => void) {
    return this.__PROMISE_SYMBOL__.then(cb, reject)
  }

  on<Key extends keyof Events>(type: Key, handler: (arg: Events[Key]) => void): void {
    this.emitter.on(type, handler)
    this.onUnmount(() => this.emitter.off(type, handler))
  }

  emit<Key extends keyof Events>(type: Key, param: Events[Key]) {
    this.emitter.emit(type, param)
  }

  off<Key extends keyof Events>(type: Key, handler?: (arg: Events[Key]) => void): void {
    this.emitter.off(type, handler)
  }

  protected onUnmount(cb: () => void) {
    this.__UNMOUNT_SYMBOL__.push(cb)
  }

  @Unmount
  // @ts-expect-error for internal
  private _unmount() {
    return Promise.all(this.__UNMOUNT_SYMBOL__.map(fn => fn()))
  }
}
