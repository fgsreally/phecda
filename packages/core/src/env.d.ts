
declare var __DEV__: boolean
declare var __PS_HMR__:((files:string[])=>void)[]