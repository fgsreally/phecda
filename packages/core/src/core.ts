import type { Phecda, PhecdaHandler } from './types'
import { mergeOptions } from './utils'

export function isPhecda(target: any) {
  return target && !!target.prototype._namespace
}
// 一个类挂载第一个phecda装饰器时，会创建对应的，类似元数据的东西
export function init(target: Phecda) {
  if (!target._namespace) {
    target._namespace = {

      /**
         * @Init 事件,
         * 引入Model时就会执行
        */
      __INIT_EVENT__: new Set(),
      /**
         * 暴露的变量，
         * 只要属性上存在至少一个装饰器，该属性就会被捕捉到
        */
      __EXPOSE_VAR__: new Set(),
      /**
         * @Ignore 绑定的属性，
         * 某属性即使被捕捉，可被强行忽略，优先级最高
        */
      __IGNORE_VAR__: new Set(),
      /**
         * 存在状态的变量
         * @deprecated
        */
      __STATE_VAR__: new Set(),
      /**
         * 状态变量的处理器
        */
      __STATE_HANDLER__: new Map(),
      /**
         * 状态变量的共有状态
        */
      __STATE_NAMESPACE__: new Map(),
    }
  }
}

export function regisInitEvent(target: Phecda, key: string) {
  init(target)
  target._namespace.__INIT_EVENT__.add(key)
}

export function getInitEvent(target: Phecda) {
  init(target)
  return [...target._namespace.__INIT_EVENT__] as string[]
}

export function setModalVar(target: Phecda, key: PropertyKey) {
  init(target)
  target._namespace.__STATE_VAR__.add(key)
  // 绑定状态的值，均属于暴露的值
  setExposeKey(target, key)
}

export function setExposeKey(target: Phecda, key: PropertyKey) {
  init(target)
  target._namespace.__EXPOSE_VAR__.add(key)
}

export function setIgnoreKey(target: Phecda, key: PropertyKey) {
  init(target)
  target._namespace.__IGNORE_VAR__.add(key)
}
// 存在状态的属性
export function getModelState(target: Phecda) {
  init(target)
  return [...target._namespace.__STATE_VAR__] as string[]
}
// 暴露的属性
// 存在状态必然暴露，反之未必，但expose可以被ignore，前者不行
// 一般而言用这个就行，某些特定情况，可用前一种
export function getExposeKey(target: Phecda) {
  init(target)

  return [...target._namespace.__EXPOSE_VAR__].filter(item => !target._namespace.__IGNORE_VAR__.has(item)) as string[]
}

export function getIgnoreKey(target: Phecda) {
  init(target)
  return [...target._namespace.__IGNORE_VAR__] as string[]
}

export function regisHandler(target: Phecda, key: PropertyKey, handler: PhecdaHandler) {
  init(target)
  if (!target._namespace.__STATE_HANDLER__.has(key))
    target._namespace.__STATE_HANDLER__.set(key, [handler])

  else
    target._namespace.__STATE_HANDLER__.get(key)!.push(handler)
}

export function getHandler(target: Phecda, key: PropertyKey) {
  return target._namespace.__STATE_HANDLER__.get(key) || []
}

export function mergeState(target: Phecda, key: PropertyKey, state: any) {
  const namespace = target._namespace.__STATE_NAMESPACE__
  if (!namespace.has(key))
    namespace.set(key, state)
  else mergeOptions(namespace.get(key)!, state)
}

export function getState(target: Phecda, key: PropertyKey) {
  const namespace = target._namespace.__STATE_NAMESPACE__
  if (namespace)
    return namespace.get(key)
}

// work for init
export function register(instance: Phecda) {
  const stateVars = getExposeKey(instance) as PropertyKey[]

  for (const item of stateVars) {
    const handlers = getHandler(instance, item)

    for (const hanlder of handlers)
      hanlder.init?.(instance)
  }
}

export async function registerAsync(instance: Phecda) {
  const stateVars = getExposeKey(instance) as PropertyKey[]

  for (const item of stateVars) {
    const handlers = getHandler(instance, item)
    for (const hanlder of handlers)
      await hanlder.init?.(instance)
  }
}
