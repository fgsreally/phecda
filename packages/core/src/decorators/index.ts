export * from './core'
export * from './function'
export * from './rule'
