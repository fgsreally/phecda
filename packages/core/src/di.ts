import type { InjectData } from './types'

// 用于自定义的情况，但不太推荐
export const DataMap = {} as InjectData

export function Provide<K extends keyof InjectData>(key: K, value: InjectData[K]) {
  DataMap[key] = value
}

export function Inject<K extends keyof InjectData>(key: K): InjectData[K] {
  return DataMap[key]
}

// 用于内部未实现的装饰器，如Watcher、Storage
export const activeInstance: Record<string, any> = {}
export function setInject(key: string, value: any) {
  activeInstance[key] = value
  return activeInstance
}

export function getInject(key: string) {
  return activeInstance[key]
}
