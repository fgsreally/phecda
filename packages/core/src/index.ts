export * from './decorators'
export * from './types'
export * from './helper'
export * from './utils'
export * from './core'
export * from './namespace'
export * from './custom/decorator'
export * from './preset'
