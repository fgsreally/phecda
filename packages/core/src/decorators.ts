import type { to } from './helper'
import { init, regisHandler, setExposeKey, setIgnoreKey, setModalVar } from './core'

export function Init(target: any, key: PropertyKey) {
  setModalVar(target, key)

  regisHandler(target, key, {
    async init(instance: any) {
      instance[key]()
    },
  })
}

export function Rule(rule: RegExp | string | Function | number,
  info: string,
  meta?: any) {
  return (obj: any, key: PropertyKey) => {
    setModalVar(obj, key)
    regisHandler(obj, key, {
      rule,
      info,
      meta,
    })
  }
}

export function Ignore(target: any, key: PropertyKey) {
  setIgnoreKey(target, key)
}

export function Clear(target: any, key: PropertyKey) {
  init(target)

  target._namespace.__INIT_EVENT__.delete(key)
  target._namespace.__EXPOSE_VAR__.delete(key)
  target._namespace.__IGNORE_VAR__.delete(key)
  target._namespace.__STATE_VAR__.delete(key)
  target._namespace.__STATE_HANDLER__.delete(key)
  target._namespace.__STATE_NAMESPACE__.delete(key)
}

export function Err<Fn extends (...args: any) => any>(cb: Fn) {
  return (target: any, key: PropertyKey) => {
    setModalVar(target, key)
    regisHandler(target, key, {
      error: cb,
    })
  }
}

export function Expose(target: any, key: PropertyKey) {
  setExposeKey(target, key)
}

export function Pipe(v: ReturnType<typeof to>) {
  return (obj: any, key: PropertyKey) => {
    setModalVar(obj, key)
    regisHandler(obj, key, {
      async pipe(instance: any) {
        const tasks = v.value
        for (const task of tasks)
          instance[key] = await task(instance[key])
      },
    })
  }
}

export function Tag(tag: string) {
  return (target: any) => {
    target.prototype.__TAG__ = tag
  }
}
// async assign value to instance
export function Assign(cb: (instance?: any) => any) {
  return (target: any) => {
    init(target.prototype)
    setModalVar(target.prototype, '__CLASS')
    regisHandler(target.prototype, '__CLASS', {
      init: async (instance: any) => {
        const value = await cb(instance)
        if (value && typeof value === 'object' && !Array.isArray(value)) {
          for (const i in value)
            instance[i] = value[i]
        }
      },
    })
  }
}

export function Global(target: any) {
  if (!(globalThis as any).__PHECDA__)
    (globalThis as any).__PHECDA__ = {}
  const tag = target.prototype.__TAG__
  if (tag)
    (globalThis as any).__PHECDA__[tag] = target
}
