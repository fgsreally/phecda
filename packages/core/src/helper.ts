/* eslint-disable new-cap */
import { SHARE_KEY, get, getExposeKey, getHandler, getState, getStateKey } from './core'
import type { AbConstruct, ClassValue, Construct, Phecda } from './types'

export function getTag<M extends Construct | AbConstruct>(moduleOrInstance: M | InstanceType<M>): PropertyKey {
  if (typeof moduleOrInstance === 'object')
    moduleOrInstance = (moduleOrInstance as InstanceType<M>).constructor

  return get(moduleOrInstance.prototype, 'tag') || (moduleOrInstance as M).name
}

export function getBind<M extends Construct | AbConstruct>(model: M) {
  // @ts-expect-error just get bind value
  const instance = new model() as Phecda
  const keys = getStateKey(instance) as PropertyKey[]
  const ret: any = {}
  for (const item of keys) {
    const state = getState(instance as any, item) as any

    if (state.value)
      ret[item] = state.value
  }
  return ret
}

export function plainToClass<M extends Construct, Data extends Record<PropertyKey, any>>(model: M, input: Data) {
  const instance: InstanceType<M> = new model()

  const keys = getExposeKey(instance) as PropertyKey[]
  for (const item of keys)

    instance[item] = input[item]

  return instance
}

// Actually not a good way
export function transformInstance<M extends Construct>(instance: InstanceType<M>, force = false) {
  const err: string[] = []
  const keys = getExposeKey(instance) as PropertyKey[]
  const addError = err.push.bind(err)

  for (const item of keys) {
    const handlers = getHandler(instance, item)
    if (handlers) {
      for (const handler of handlers) {
        const pipe = handler.pipe
        if (!pipe)
          continue
        pipe(instance, addError)

        if (err.length && !force)
          return err
      }
    }
  }
  return err
}

export async function transformInstanceAsync<M extends Construct>(instance: InstanceType<M>, force = false) {
  const err: string[] = []
  const keys = getExposeKey(instance) as PropertyKey[]
  const addError = err.push.bind(err)

  for (const item of keys) {
    const handlers = getHandler(instance, item)

    if (handlers) {
      for (const handler of handlers) {
        const pipe = handler.pipe
        if (!pipe)
          continue
        await pipe(instance, addError)

        if (err.length && !force)
          return err
      }
    }
  }
  return err
}

export function transformProperty<M extends Construct>(instance: InstanceType<M>, property: keyof InstanceType<M>, force = false) {
  const err: string[] = []
  const handlers = getHandler(instance, property)
  const addError = err.push.bind(err)
  if (handlers) {
    for (const handler of handlers) {
      const pipe = handler.pipe
      if (!pipe)
        continue

      pipe(instance, addError)

      if (err.length && !force)
        return err
    }
  }
  return err
}

export async function transformPropertyAsync<M extends Construct>(instance: InstanceType<M>, property: keyof InstanceType<M>, force = false) {
  const err: string[] = []
  const handlers = getHandler(instance, property)
  const addError = err.push.bind(err)
  if (handlers) {
    for (const handler of handlers) {
      const pipe = handler.pipe
      if (!pipe)
        continue

      await pipe(instance, addError)

      if (err.length && !force)
        return err
    }
  }
  return err
}

export function classToPlain<M>(instance: M): ClassValue<M> {
  const data = {} as any
  const exposeVars = getExposeKey(instance as any) as PropertyKey[]
  for (const item of exposeVars)
    data[item] = (instance as any)[item]

  return JSON.parse(JSON.stringify(data))
}

export function snapShot<T extends Construct>(data: InstanceType<T>) {
  const snap = {} as unknown as InstanceType<T>
  for (const i in data)
    snap[i] = data[i]

  return {
    data,
    clear() {
      for (const i in snap)
        delete data[i]
    },
    apply() {
      for (const i in snap)
        data[i] = snap[i]
    },
  }
}
/**
 * add decorator to a class by function
 */
export function addDecoToClass<M extends Construct | AbConstruct>(c: M, key: keyof InstanceType<M> | PropertyKey, handler: PropertyDecorator | ClassDecorator) {
  handler(key === SHARE_KEY ? c : c.prototype, key as any)
}

export function Pipeline(...decos: ((...args: any) => void)[]) {
  return (...args: any) => {
    for (const d of decos)
      d(...args)
  }
}

export function isAsyncFunc(fn: Function) {
  return (fn as any)[Symbol.toStringTag] === 'AsyncFunction'
}
