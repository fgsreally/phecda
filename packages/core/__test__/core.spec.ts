import { describe, expect, it } from 'vitest'
import { getExposeKey, getOwnExposeKey, getState, setModelVar, setState } from '../src'

describe('extends won\'t populate namespace', () => {
  it('modelVar', () => {
    function Test(target: any, key: any) {
      setModelVar(target, key)
    }
    class A {
      @Test
      a: string
    }

    class B extends A {
      @Test
        b: string
    }

    class C extends A {
      @Test
        c: string
    }
    const instanceB: any = new B()
    const instanceC: any = new C()

    expect(getExposeKey(instanceB)).toMatchSnapshot()
    expect(getOwnExposeKey(instanceB)).toMatchSnapshot()
    expect(getExposeKey(instanceC)).toMatchSnapshot()
    expect(getOwnExposeKey(instanceC)).toMatchSnapshot()
  })

  it('setState', () => {
    function Test(value: any) {
      return (target: any, key: any) => {
        setState(target, key, value)
      }
    }
    class A {
      @Test({ a: 1, b: 0 })
      x: string
    }

    class B extends A {
      @Test({ b: 1 })
        x: string
    }

    class C extends A {
      @Test({ c: 1 })
        x: string
    }

    const instanceB = new B()
    const instanceC = new C()
    expect(getState(instanceB as any, 'x')).toMatchSnapshot()
    expect(getState(instanceC as any, 'x')).toMatchSnapshot()
  })
})
