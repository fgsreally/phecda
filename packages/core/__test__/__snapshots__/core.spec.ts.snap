// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`extends won't populate namespace > expose keys 1`] = `
[
  "b",
  "a",
]
`;

exports[`extends won't populate namespace > expose keys 2`] = `
[
  "b",
]
`;

exports[`extends won't populate namespace > expose keys 3`] = `
[
  "c",
  "a",
]
`;

exports[`extends won't populate namespace > expose keys 4`] = `
[
  "c",
]
`;

exports[`extends won't populate namespace > expose keys 5`] = `
[
  "a",
]
`;

exports[`extends won't populate namespace > expose keys 6`] = `
[
  "c",
]
`;

exports[`extends won't populate namespace > expose keys 7`] = `
[
  "e",
  "c",
  "a",
]
`;

exports[`extends won't populate namespace > expose keys 8`] = `
[
  "e",
]
`;

exports[`extends won't populate namespace > setState 1`] = `
{
  "a": 1,
  "b": 1,
}
`;

exports[`extends won't populate namespace > setState 2`] = `
{
  "a": 1,
  "b": 0,
  "c": 1,
}
`;
